### DISCUSSION TITLE
Documentation on various timeouts

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [pdezwart/php-amqp#505] [Q&A] [answered]
Hello, 

We are looking into configuration tuning for our AMQP setup. We see there are various timeouts possible:
- amqp.timeout
- amqp.confirm_timeout
- amqp.connect_timeout
- amqp.read_timeout
- amqp.write_timeout

However, we are unable to find a clear explanation as to what these timeouts refer. Does anyone with this knowledge would share it with us? 

This could go into a doc PR later on :) 

Our usecase is:

- we find that at times the `AMQPQueue->get()` operation gets stuck indefinitely without throwing an error. 
- we tried setting the `amqp.read_timeout` to 5s, as a get is a read operation, but this did not solve the issue
- we tried setting the `amqp.write_timeout` to 5s too, but this did not solve the issue.

Going in blind prevents us from easily spotting & fixing this issue :) 

Thank you!

**Comment by lstrojny** [comment, marked as answer]
First, we have a couple of timeouts that are configured on the `AMQPConnection` object:

 * `connect_timeout`: TCP level timeout for connecting to the AMQP server. What makes sense here depends on the reliability of your network.
 * `read_timeout`: how long does `AMQPQueue::consume` wait until it times out
 * `write_timeout`: how long does `AMQPExchange::publish` wait until it times out
 * `rpc_timeout`: timeout for sending the actual command to the AMQP server
 * `timeout`: deprecated, for BC reasons does the same as setting the read timeout

`AMQPQueue::get` is not bound to read timeout, instead it will return null immediately if no message is available. `AMQPQueue::consume` respects the read timeout. Always use `AMQPQueue::consume` to consume a ton of messages, it offers significantly more message throughput.

`confirm_timeout` does not exist as a connection parameter but instead there is `AMQPChannel::waitForConfirm()` where one can specify a timeout (for [publisher confirms](https://www.rabbitmq.com/confirms.html))
